# Lab 2  - Exercises 4  5
## Modified by: Santiago Carvajal

### SPAM Detection - The Naive Bayes Algorithm in Python with Scikit-Learn 
D. Shahrokhian
https://stackabuse.com/the-naive-bayes-algorithm-in-python-with-scikit-learn/

In [1]:
import pandas as pd

# Reading data from Table 13.1
df_131 = pd.read_table('Tabla131',  
                   sep='\t', 
                   header=None,
                   names=['label', 'message'])
# Reading data from Table 13.1
df_1310 = pd.read_table('Tabla1310',  
                   sep='\t', 
                   header=None,
                   names=['label', 'message'])
# Reading data from IAML
df_iaml = pd.read_table('IAML',  
                   sep='\t', 
                   header=None,
                   names=['label', 'message'])
print("Table 13.1")
print(df_131.head())
print("\nTable 13.10")
print(df_1310.head())
print("\nIAML Data")
print(df_iaml.head())

Table 13.1
  label                              message
0   yes              Chinese Beijing Chinese
1   yes             Chinese Chinese Shanghai
2   yes                        Chinese Macao
3    no                  Tokyo Japan Chinese
4   yes  Chinese Chinese Chinese Tokyo Japan

Table 13.10
  label                message
0   yes         Taipei Taiwan 
1   yes  Macao Taiwan Shanghai
2    no           apan Sapporo
3    no   Sapporo Osaka Taiwan
4    no  Taiwan Taiwan Sapporo

IAML Data
  label                message
0  spam  send us your password
1   ham    send us your review
2   ham   review your password
3  spam              review us
4  spam     send your password


In [2]:
df_131['label'] = df_131.label.map({'no': 0, 'yes': 1})
df_131['message'] = df_131.message.map(lambda x: x.lower())
df_131['message'] = df_131.message.str.replace('[^\w\s]', '')
print("Table 13.1")
print(df_131.head())

df_1310['label'] = df_1310.label.map({'no': 0, 'yes': 1})
df_1310['message'] = df_1310.message.map(lambda x: x.lower())
df_1310['message'] = df_1310.message.str.replace('[^\w\s]', '')
print("\nTable 13.10")
print(df_1310.head())

df_iaml['label'] = df_iaml.label.map({'ham': 0, 'spam': 1})
df_iaml['message'] = df_iaml.message.map(lambda x: x.lower())
df_iaml['message'] = df_iaml.message.str.replace('[^\w\s]', '')
print("\nIAML data")
print(df_iaml.head())

Table 13.1
   label                              message
0      1              chinese beijing chinese
1      1             chinese chinese shanghai
2      1                        chinese macao
3      0                  tokyo japan chinese
4      1  chinese chinese chinese tokyo japan

Table 13.10
   label                message
0      1         taipei taiwan 
1      1  macao taiwan shanghai
2      0           apan sapporo
3      0   sapporo osaka taiwan
4      0  taiwan taiwan sapporo

IAML data
   label                message
0      1  send us your password
1      0    send us your review
2      0   review your password
3      1              review us
4      1     send your password


In [3]:
# https://www.nltk.org/ Natural Language Toolkit
# Punkt Sentence Tokenizer https://www.nltk.org/api/nltk.tokenize.html#module-nltk.tokenize.punkt
import nltk
nltk.download('punkt')

df_131['message'] = df_131['message'].apply(nltk.word_tokenize)
print("Table 13.1")
print(df_131.head())
df_1310['message'] = df_1310['message'].apply(nltk.word_tokenize)
print("\nTable 13.10")
print(df_1310.head())
df_iaml['message'] = df_iaml['message'].apply(nltk.word_tokenize)
print("\nIAML data")
print(df_iaml.head())

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Santiago\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
Table 13.1
   label                                    message
0      1                [chinese, beijing, chinese]
1      1               [chinese, chinese, shanghai]
2      1                           [chinese, macao]
3      0                    [tokyo, japan, chinese]
4      1  [chinese, chinese, chinese, tokyo, japan]

Table 13.10
   label                    message
0      1           [taipei, taiwan]
1      1  [macao, taiwan, shanghai]
2      0            [apan, sapporo]
3      0   [sapporo, osaka, taiwan]
4      0  [taiwan, taiwan, sapporo]

IAML data
   label                     message
0      1  [send, us, your, password]
1      0    [send, us, your, review]
2      0    [review, your, password]
3      1                [review, us]
4      1      [send, your, password]


In [4]:
# https://www.nltk.org/api/nltk.stem.html
#https://tartarus.org/martin/PorterStemmer/
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()
df_131['message'] = df_131['message'].apply(lambda x: [stemmer.stem(y) for y in x]) 
print("Table 13.1")
print(df_131.head())
df_1310['message'] = df_1310['message'].apply(lambda x: [stemmer.stem(y) for y in x]) 
print("\nTable 13.10")
print(df_1310.head())
df_iaml['message'] = df_iaml['message'].apply(lambda x: [stemmer.stem(y) for y in x]) 
print("\nIAML data")
print(df_iaml.head())

Table 13.1
   label                                 message
0      1                  [chines, beij, chines]
1      1              [chines, chines, shanghai]
2      1                         [chines, macao]
3      0                  [tokyo, japan, chines]
4      1  [chines, chines, chines, tokyo, japan]

Table 13.10
   label                    message
0      1           [taipei, taiwan]
1      1  [macao, taiwan, shanghai]
2      0            [apan, sapporo]
3      0   [sapporo, osaka, taiwan]
4      0  [taiwan, taiwan, sapporo]

IAML data
   label                     message
0      1  [send, us, your, password]
1      0    [send, us, your, review]
2      0    [review, your, password]
3      1                [review, us]
4      1      [send, your, password]


In [5]:
# Converts the list of words into space-separated strings
df_131['message'] = df_131['message'].apply(lambda x: ' '.join(x))
print("Table 13.1")
print(df_131.head())
df_1310['message'] = df_1310['message'].apply(lambda x: ' '.join(x))
print("\nTable 13.10")
print(df_1310.head())
df_iaml['message'] = df_iaml['message'].apply(lambda x: ' '.join(x))
print("\nIAML data")
print(df_iaml.head())


Table 13.1
   label                           message
0      1                chines beij chines
1      1            chines chines shanghai
2      1                      chines macao
3      0                tokyo japan chines
4      1  chines chines chines tokyo japan

Table 13.10
   label                message
0      1          taipei taiwan
1      1  macao taiwan shanghai
2      0           apan sapporo
3      0   sapporo osaka taiwan
4      0  taiwan taiwan sapporo

IAML data
   label                message
0      1  send us your password
1      0    send us your review
2      0   review your password
3      1              review us
4      1     send your password


In [6]:
# Convert a collection of text documents to a matrix of token counts
# http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
# http://scikit-learn.org/stable/modules/feature_extraction.html#text-feature-extraction
# to allow one letter words count_vect = CountVectorizer(token_pattern = r"(?u)\b\w+\b")
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer(token_pattern = r"(?u)\b\w+\b") 
counts_131 = count_vect.fit_transform(df_131['message'])  
counts_1310 = count_vect.fit_transform(df_1310['message']) 
counts_iaml = count_vect.fit_transform(df_iaml['message']) 

print("--------Table 13.1 data------------------------------------")
print(counts_131)
print("--------Table 13.10 data------------------------------------")
print(counts_1310)
print("--------IAML data------------------------------------")
print(counts_iaml)

--------Table 13.1 data------------------------------------
  (0, 0)	1
  (0, 1)	2
  (1, 4)	1
  (1, 1)	2
  (2, 3)	1
  (2, 1)	1
  (3, 2)	1
  (3, 5)	1
  (3, 1)	1
  (4, 2)	1
  (4, 5)	1
  (4, 1)	3
--------Table 13.10 data------------------------------------
  (0, 6)	1
  (0, 5)	1
  (1, 4)	1
  (1, 1)	1
  (1, 6)	1
  (2, 3)	1
  (2, 0)	1
  (3, 2)	1
  (3, 3)	1
  (3, 6)	1
  (4, 3)	1
  (4, 6)	2
--------IAML data------------------------------------
  (0, 2)	1
  (0, 6)	1
  (0, 5)	1
  (0, 4)	1
  (1, 3)	1
  (1, 6)	1
  (1, 5)	1
  (1, 4)	1
  (2, 3)	1
  (2, 2)	1
  (2, 6)	1
  (3, 3)	1
  (3, 5)	1
  (4, 2)	1
  (4, 6)	1
  (4, 4)	1
  (5, 0)	1
  (5, 6)	1
  (5, 5)	1
  (5, 4)	1
  (6, 1)	1
  (6, 3)	1
  (6, 5)	1


In [7]:
print(counts_131.shape)
print(counts_1310.shape)
print(counts_iaml.shape)

(5, 6)
(5, 7)
(7, 7)


## Mutlinomial Naive Bayes Classification with sklearn ##

### Classification with Table 13.1 data ###

In [8]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with Table 13.1 data----")
X_train, X_test, y_train, y_test = train_test_split(counts_131, df_131['label'], test_size=0.2, shuffle = False ,random_state=69) 

model = MultinomialNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with Table 13.1 data----
Probabilities for each class 
[[0.31024139 0.68975861]]
Accuracy: (0 - 1)
1.0
Confusion matrix: 
[[1]]


### Classification with Table 13.10 data ###

In [9]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with Table 13.10 data----")
X_train, X_test, y_train, y_test = train_test_split(counts_1310, df_1310['label'], test_size=0.2, shuffle = False ,random_state=69) 

model = MultinomialNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with Table 13.10 data----
Probabilities for each class 
[[0.57142857 0.42857143]]
Accuracy: (0 - 1)
1.0
Confusion matrix: 
[[1]]


### Classification with IAML data ###

In [10]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with IAML data----")
X_train, X_test, y_train, y_test = train_test_split(counts_iaml, df_iaml['label'], test_size=0.1, shuffle = False ,random_state=69) 

model = MultinomialNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with IAML data----
Probabilities for each class 
[[0.52228412 0.47771588]]
Accuracy: (0 - 1)
1.0
Confusion matrix: 
[[1]]


### Analysis ###
We get different probabilities using the sklearn Naive Bayes library because they perform a normalization proccess before returning the final results of the probability of each class. Thats why we see bigger probabilities values

## Punto 5: Bernoulli Naive Bayes Classification with sklearn ##

### Classification with Table 13.1 data ###

In [11]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with Table 13.1 data----")
X_train, X_test, y_train, y_test = train_test_split(counts_131, df_131['label'], test_size=0.2, shuffle = False ,random_state=69) 

model = BernoulliNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with Table 13.1 data----
Probabilities for each class 
[[0.80893321 0.19106679]]
Accuracy: (0 - 1)
0.0
Confusion matrix: 
[[0 0]
 [1 0]]


### Classification with Table 13.10 data ###

In [12]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with Table 13.10 data----")
X_train, X_test, y_train, y_test = train_test_split(counts_1310, df_1310['label'], test_size=0.2, shuffle = False ,random_state=69) 

model = BernoulliNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with Table 13.10 data----
Probabilities for each class 
[[0.75 0.25]]
Accuracy: (0 - 1)
1.0
Confusion matrix: 
[[1]]


### Classification with IMAL data ###

In [13]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
import numpy as np

#Classiification with Table 13.1 data
print("----Classification with IAML data----")
X_train, X_test, y_train, y_test = train_test_split(counts_iaml, df_iaml['label'], test_size=0.1, shuffle = False ,random_state=69) 

model = BernoulliNB().fit(X_train, y_train)  

predicted = model.predict(X_test)
predicted_proba = model.predict_proba(X_test)
print("Probabilities for each class ")
print(predicted_proba)
print("Accuracy: (0 - 1)")
print(np.mean(predicted == y_test))
from sklearn.metrics import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(y_test, predicted))

----Classification with IAML data----
Probabilities for each class 
[[0.61565168 0.38434832]]
Accuracy: (0 - 1)
1.0
Confusion matrix: 
[[1]]


### Analysis ###

| Data Table | MultinomialResult | BernoulliResult |
| --- | --- | --- |
| Table 13.1 | [0.31024139 0.68975861] | [0.80893321 0.19106679]|
| Table 13.10 | [0.57142857 0.42857143] | [0.75 0.25]|
| IAML data | [0.52228412 0.47771588] | [0.61565168 0.38434832]|




As we can see in previous Table, we obtained similar results for data in Table 13.10 with Bernoulli and Multinomial Classificator.

But we get the opposite results using data in Table 13.1 with the differents classificators. This result is consistent with the exercises that we did by hand. 

Using IAML data, we obtained similar results with both classificators.

As we mention before, that probabilities are not the same because of the normalization that the library we are using does.



